docs/developing.md

```diff
@@ -0,0 +1,309 @@
+    Recording the "versions" of a SMIRNOFF spec tag allows us to encode the default behavior and API of a specific generation of ParameterHandlers, while allowing the safe addition of new attributes and behaviors.
+
+  Each ParameterHandler-derived class MAY implement:
+    - `known_kwargs`: Keyword arguments passed to `ForceField.create_openmm_system` are validated against the `known_kwargs` lists of each ParameterHandler that the ForceField owns.
+      If present, these kwargs and their values will be passed on to the ParameterHandler.
+    - `to_dict`: converts the ParameterHandler to a hierarchical dict compliant with the SMIRNOFF specification.
+      The default implementation of this function should suffice for most developers.
+    - `check_handler_compatibility`: Checks whether this ParameterHandler is "compatible" with another.
+      This function is used when a ForceField is attempted to be constructed from *multiple* SMIRNOFF data sources, and it is necessary to check that two sections with the same tagname can be combined in a sane way.
+      For example, if the user instructed two `vdW` sections to be read, but the sections defined different vdW potentials, then this function should raise an Exception indicating that there is no safe way to combine the parameters.
+      The default implementation of this function should suffice for most developers.
+    - `postprocess_system`: operates identically to `create_force`, but is run after each ParameterHandlers' `create_force` has already been called.
+      The default implementation of this method simply does nothing, and should suffice for most developers.
+
+[`ParameterAttribute`](openff.toolkit.typing.engines.smirnoff.parameters.ParameterAttribute)
+: A single parameter value with units and runtime validation.
+
+[`IndexedParameterAttribute`](openff.toolkit.typing.engines.smirnoff.parameters.IndexedParameterAttribute)
+: A variable number of parameter values with units and runtime validation.
+For example, dihedral torsions are often parameterized as the sum of several sine wave terms.
+Each of the parameters of the sine wave `k`, `periodicity`, and `phase` are implemented as `IndexedParameterAttribute`s.
+
+[`ParameterType`](openff.toolkit.typing.engines.smirnoff.parameters.ParameterType)
+: A base class for the parameters of a `ParameterHandler`.
+  Extend this alongside `ParameterHandler` to define and validate the parameters of a new force.
+  For example, the Lennard-Jones potential can be parameterized through either the length `ParameterAttribute` `sigma` or `r_min`, alongside the energy `ParameterAttribute` `epsilon`. Both options are handled through the [`vdWType`](openff.toolkit.typing.engines.smirnoff.parameters.vdWHandler.vdWType) class, a subclass of `ParameterType`.
+  This is distinct from the [`vdWHandler`](openff.toolkit.typing.engines.smirnoff.parameters.vdWHandler) class, which includes other `ParameterAttribute`s that apply to all atoms in the system.
```

Comment:
```suggestion
[`ParameterType`](openff.toolkit.typing.engines.smirnoff.parameters.ParameterType)
: A base class for the SMIRKS-based parameters of a `ParameterHandler`.
  Extend this alongside `ParameterHandler` to define and validate the SMIRKS-based (/ForceField) parameters of a new force.
  This is analagous to ParmEd's XType classes, like [BondType](https://parmed.github.io/ParmEd/html/api/parmed/parmed.html?highlight=bondtype#parmed.BondType). 
  
  For example, the Lennard-Jones potential can be parameterized through either the size `ParameterAttribute` `sigma` or `r_min`, alongside the energy `ParameterAttribute` `epsilon`. Both options are handled through the [`vdWType`](openff.toolkit.typing.engines.smirnoff.parameters.vdWHandler.vdWType) class, a subclass of `ParameterType`.
  
  
  
  ~This is distinct from the [`vdWHandler`](openff.toolkit.typing.engines.smirnoff.parameters.vdWHandler) class, which includes other `ParameterAttribute`s that apply to all atoms in the system.~ Somewhere we should make the statement that BOTH ParameterTypes AND ParameterHandlers should be defining ParameterAttributes
```

Reply:
Incorporated these suggestions in dbcc59a